File: reachables/src/main/java/act/installer/reachablesexplorer/Reachable.java
/*************************************************************************
*                                                                        *
*  This file is part of the 20n/act project.                             *
*  20n/act enables DNA prediction for synthetic biology/bioengineering.  *
*  Copyright (C) 2017 20n Labs, Inc.                                     *
*                                                                        *
*  Please direct all queries to act@20n.com.                             *
*                                                                        *
*  This program is free software: you can redistribute it and/or modify  *
*  it under the terms of the GNU General Public License as published by  *
*  the Free Software Foundation, either version 3 of the License, or     *
*  (at your option) any later version.                                   *
*                                                                        *
*  This program is distributed in the hope that it will be useful,       *
*  but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*  GNU General Public License for more details.                          *
*                                                                        *
*  You should have received a copy of the GNU General Public License     *
*  along with this program.  If not, see <http://www.gnu.org/licenses/>. *
*                                                                        *
*************************************************************************/

package act.installer.reachablesexplorer;


import act.shared.Chemical;
import act.shared.helpers.MongoDBToJSON;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.mongodb.BasicDBObject;
import com.mongodb.DBObject;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@JsonInclude(JsonInclude.Include.ALWAYS)
public class Reachable {
  @JsonIgnore
  static final String ID_FIELD_NAME = "_id";
  @JsonIgnore
  static final String INCHI_FIELD_NAME = "InChI";


  @JsonProperty("_id")
  private Long id;
  private String pageName;
  private String pathwayVisualization;
  private Boolean isNative;
  @JsonProperty("InChI")
  private String inchi;
  private String smiles;
  private String inchiKey;
  private String structureFilename;
  private List<String> names;
  private String wordCloudFilename;
  private PrecursorData precursorData;
  private Map<Chemical.REFS, BasicDBObject> xref;
  private SynonymData synonyms;
  private List<PatentSummary> patentSummaries;
  private PhysiochemicalProperties physiochemicalProperties;

  public Reachable() {}

  public Reachable(Long id,
                   String pageName,
                   String inchi,
                   String smiles,
                   String inchiKey,
                   List<String> names,
                   PrecursorData precursors,
                   SynonymData synonyms,
                   Boolean isNative,
                   String structureFilename,
                   String wordCloudFilename,
                   String pathwayVisualization,
                   Map<Chemical.REFS, BasicDBObject> xref,
                   List<PatentSummary> patentSummaries,
                   PhysiochemicalProperties physiochemicalProperties) {
    this.id = id;
    this.pageName = pageName;
    this.inchi = inchi;
    this.smiles = smiles;
    this.inchiKey = inchiKey;
    this.names = names;
    this.precursorData = precursors;
    this.synonyms = synonyms;
    this.isNative = isNative;
    this.structureFilename = structureFilename;
    this.wordCloudFilename = wordCloudFilename;
    this.pathwayVisualization = pathwayVisualization;
    this.xref = xref;
    this.patentSummaries = patentSummaries;
    this.physiochemicalProperties = physiochemicalProperties;
  }

  public Reachable(
      Long id, String pageName, String inchi, String smiles, String inchiKey, List<String> names,
      SynonymData synonymData, String structureFilename, String wordCloudFilename, Map<Chemical.REFS,
      BasicDBObject> xref, PhysiochemicalProperties physiochemicalProperties) {
    // TODO: these arguments are out of hand, and this could use a Builder.
    this(id, pageName, inchi, smiles, inchiKey, names, new PrecursorData(), synonymData, null,
        structureFilename, wordCloudFilename, null, xref, null, physiochemicalProperties);
  }

  public Long getId() {
    return id;
  }

  public String getPageName() {
    return pageName;
  }

  public void setPageName(String pageName) {
    this.pageName = pageName;
  }

  public String getPathwayVisualization() {
    return pathwayVisualization;
  }

  public void setPathwayVisualization(String pathwayVisualization) {
    this.pathwayVisualization = pathwayVisualization;
  }

  public Boolean getNative() {
    return isNative;
  }

  public void setNative(Boolean aNative) {
    isNative = aNative;
  }

  @JsonIgnore
  public String getInchi() {
    return inchi;
  }

  @JsonIgnore
  public void setInchi(String inchi) {
    this.inchi = inchi;
  }

  public String getSmiles() {
    return smiles;
  }

  public void setSmiles(String smiles) {
    this.smiles = smiles;
  }

  public String getInchiKey() {
    return inchiKey;
  }

  public void setInchiKey(String inchiKey) {
    this.inchiKey = inchiKey;
  }

  public String getStructureFilename() {
    return structureFilename;
  }

  public void setStructureFilename(String structureFilename) {
    this.structureFilename = structureFilename;
  }

  public List<String> getNames() {
    return names;
  }

  public void setNames(List<String> names) {
    this.names = names;
  }

  public String getWordCloudFilename() {
    return wordCloudFilename;
  }

  public void setWordCloudFilename(String wordCloudFilename) {
    this.wordCloudFilename = wordCloudFilename;
  }

  public PrecursorData getPrecursorData() {
    return precursorData;
  }

  public void setPrecursorData(PrecursorData precursorData) {
    this.precursorData = precursorData;
  }

  public Map<Chemical.REFS, BasicDBObject> getXref() {
    return xref;
  }

  public void setXref(Map<Chemical.REFS, BasicDBObject> xref) {
    this.xref = xref;
  }

  public SynonymData getSynonyms() {
    return synonyms;
  }

  public void setSynonyms(SynonymData synonyms) {
    this.synonyms = synonyms;
  }

  public List<PatentSummary> getPatentSummaries() {
    return this.patentSummaries;
  }

  public void setPatentSummaries(List<PatentSummary> patentSummaries) {
    this.patentSummaries = patentSummaries;
  }

  public PhysiochemicalProperties getPhysiochemicalProperties() {
    return physiochemicalProperties;
  }

  public void setPhysiochemicalProperties(PhysiochemicalProperties physiochemicalProperties) {
    this.physiochemicalProperties = physiochemicalProperties;
  }

  public void addPatentSummaries(List<PatentSummary> patentSummaries) {
    if (this.patentSummaries == null) {
      this.patentSummaries = new ArrayList<>();
    }
    this.patentSummaries.addAll(patentSummaries);
  }

}

File: reachables/src/main/java/act/installer/reachablesexplorer/Reachable.scala
/*package act.installer.reachablesexplorer

import act.shared.Chemical
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.mongodb.BasicDBObject
import java.util.ArrayList
import java.util.List
import java.util.Map

@JsonInclude(JsonInclude.Include.ALWAYS) object Reachable {
  @JsonIgnore private[reachablesexplorer] val inchiFieldName: String = "InChI"
}

@JsonInclude(JsonInclude.Include.ALWAYS) class Reachable {
  @JsonProperty("_id") private var id: Long = 0L
  private var pageName: String = null
  private var pathwayVisualization: String = null
  private var isNative: Boolean = false
  @JsonProperty("InChI") private var inchi: String = null
  private var smiles: String = null
  private var inchiKey: String = null
  private var structureFilename: String = null
  private var names: util.List[String] = null
  private var wordCloudFilename: String = null
  private var precursorData: PrecursorData = null
  private var xref: util.Map[Chemical.REFS, BasicDBObject] = null
  private var synonyms: SynonymData = null
  private var patentSummaries: util.List[PatentSummary] = null

  def this(id: Long, pageName: String, inchi: String, smiles: String, inchiKey: String, names: util.List[String], precursors: PrecursorData, synonyms: SynonymData, isNative: Boolean, structureFilename: String, wordCloudFilename: String, pathwayVisualization: String, xref: util.Map[Chemical.REFS, BasicDBObject], patentSummaries: util.List[PatentSummary]) {
    this()
    this.id = id
    this.pageName = pageName
    this.inchi = inchi
    this.smiles = smiles
    this.inchiKey = inchiKey
    this.names = names
    this.precursorData = precursors
    this.synonyms = synonyms
    this.isNative = isNative
    this.structureFilename = structureFilename
    this.wordCloudFilename = wordCloudFilename
    this.pathwayVisualization = pathwayVisualization
    this.xref = xref
    this.patentSummaries = patentSummaries
  }

  def this(id: Long, pageName: String, inchi: String, smiles: String, inchiKey: String, names: util.List[String], synonymData: SynonymData, structureFilename: String, wordCloudFilename: String, xref: util.Map[Chemical.REFS, BasicDBObject]) {
    this(id, pageName, inchi, smiles, inchiKey, names, new PrecursorData, synonymData, null, structureFilename, wordCloudFilename, null, xref, null)
  }

  def getPageName: String = {
    return pageName
  }

  def setPageName(pageName: String) {
    this.pageName = pageName
  }

  def getPathwayVisualization: String = {
    return pathwayVisualization
  }

  def setPathwayVisualization(pathwayVisualization: String) {
    this.pathwayVisualization = pathwayVisualization
  }

  def getNative: Boolean = {
    return isNative
  }

  def setNative(aNative: Boolean) {
    isNative = aNative
  }

  @JsonIgnore def getInchi: String = {
    return inchi
  }

  @JsonIgnore def setInchi(inchi: String) {
    this.inchi = inchi
  }

  def getSmiles: String = {
    return smiles
  }

  def setSmiles(smiles: String) {
    this.smiles = smiles
  }

  def getInchiKey: String = {
    return inchiKey
  }

  def setInchiKey(inchiKey: String) {
    this.inchiKey = inchiKey
  }

  def getStructureFilename: String = {
    return structureFilename
  }

  def setStructureFilename(structureFilename: String) {
    this.structureFilename = structureFilename
  }

  def getNames: util.List[String] = {
    return names
  }

  def setNames(names: util.List[String]) {
    this.names = names
  }

  def getWordCloudFilename: String = {
    return wordCloudFilename
  }

  def setWordCloudFilename(wordCloudFilename: String) {
    this.wordCloudFilename = wordCloudFilename
  }

  def getPrecursorData: PrecursorData = {
    return precursorData
  }

  def setPrecursorData(precursorData: PrecursorData) {
    this.precursorData = precursorData
  }

  def getXref: util.Map[Chemical.REFS, BasicDBObject] = {
    return xref
  }

  def setXref(xref: util.Map[Chemical.REFS, BasicDBObject]) {
    this.xref = xref
  }

  def getSynonyms: SynonymData = {
    return synonyms
  }

  def setSynonyms(synonyms: SynonymData) {
    this.synonyms = synonyms
  }

  def getPatentSummaries: util.List[PatentSummary] = {
    return this.patentSummaries
  }

  def setPatentSummaries(patentSummaries: util.List[PatentSummary]) {
    this.patentSummaries = patentSummaries
  }

  def addPatentSummaries(patentSummaries: util.List[PatentSummary]) {
    if (this.patentSummaries == null) {
      this.patentSummaries = new util.ArrayList[PatentSummary]
    }
    this.patentSummaries.addAll(patentSummaries)
  }
}**/